synapse/lib/hiveauth.py

```diff
@@ -1248,7 +1255,22 @@ async def tryPasswd(self, passwd, nexs=True):
             return False
 
         if isinstance(shadow, dict):
-            return await s_passwd.checkShadowV2(passwd=passwd, shadow=shadow)
+            result = await s_passwd.checkShadowV2(passwd=passwd, shadow=shadow)
+
+            if self.auth.policy and (attempts := self.auth.policy['attempts']) > 0:
+                if result:
+                    await self.auth.setUserInfo(self.iden, 'policy:attempts', 0)
+                    return True
+
+                valu = self.info.get('policy:attempts') + 1
+                await self.auth.setUserInfo(self.iden, 'policy:attempts', valu)
+
+                if valu >= attempts:
+                    await self.setLocked(True, logged=nexs)
+                    mesg = f'Account locked due to exceeding maximum number of incorrect attempts ({valu + 1}).'
```

Comment:
Before we lockout the user, we'll want to log this message with the same attributes ( as close as possible ) and status code used in setUserLocked. Then we'll want to do the lockout, then we can raise the exception.

Something like:
```
logger.info(f'User {self.name} has exceeded the number of allowed password attempts ({valu +1}), locking their account.'), extra= {'synapse': {'target_user': user.iden, 'target_username': user.name, 'status': 'MODIFY'}})
```